source/base/parameter_handler.cc

```diff
@@ -584,6 +581,105 @@ void ParameterHandler::parse_input_from_xml (std::istream &in)
 
 
 
+namespace
+{
+  // Recursively go through the 'source' tree and see if we can find
+  // corresponding entries in the 'destination' tree. If not, error out
+  // (i.e. we have just read an JSON file that has entries that weren't
+  // declared in the ParameterHandler object); if so, copy the value of these
+  // nodes into the destination object
+  void
+  read_json_recursively (const boost::property_tree::ptree &source,
```

Comment:
How does this function differ from the corresponding XML one? If I see this correctly, then the input is just the ptree, and it shouldn't matter whether we got that ptree by reading XML or reading JSON?

Reply:
It doesn't differ yet. I am currently looking into supporting json arrays, and then it will differ quite a lot.

Reply:
I was wrong, it differs a tiny bit. The json actually uses mangle on the key, where the xml one doesn't. I personally think the xml could also use the mangle, but I am sure the json one needs it. So should I keep them separate, or add the mangle to the xml file? And should I change the name then to xml_json or leave it at xml?

Reply:
I'm actually pretty sure the XML also uses the mangle on the keys -- because that's how we internally store the data in `ParameterHandler`: the key is *always* mangled in the ptree data structure.

You can see that if you compare the `tests/parameter_handler_read_xml.cc` and `tests/prm/parameter_handler_read_xml.prm` files. Look, for example, for the "double 1" and "string list" parameters that have spaces in them.

So if you're mangling on the key, I wonder if you're really mangling twice.

Reply:
The declarations in `tests/parameter_handler/parameter_handler_read_xml.cc` have spaces in them, but the xml file `tests/parameter_handler/prm/parameter_handler_read_xml.prm` which is read in doesn't (see for example `double_201`):

```

<?xml version="1.0" encoding="utf-8"?>
<ParameterHandler><int1><value>2</value><default_value>1</default_value><documentation>doc 1</documentation><pattern>0</pattern><pattern_description>[Integer range -2147483648...2147483647 (inclusive)]</pattern_description></int1><int2><value>3</value><default_value>2</default_value><documentation>doc 2</documentation><pattern>1</pattern><pattern_description>[Integer range -2147483648...2147483647 (inclusive)]</pattern_description></int2><ss1><double_201><value>2.234</value><default_value>1.234</default_value><documentation>doc 3</documentation><pattern>2</pattern><pattern_description>[Double -1.79769e+308...1.79769e+308 (inclusive)]</pattern_description></double_201><ss2><double_202><value>5.321</value><default_value>4.321</default_value><documentation>doc 4</documentation><pattern>3</pattern><pattern_description>[Double -1.79769e+308...1.79769e+308 (inclusive)]</pattern_description></double_202></ss2></ss1><Testing_25testing><string_26list><value>__&lt; &amp; &gt; ; /</value><default_value>&lt; &amp; &gt; ; /</default_value><documentation>docs 1</documentation><pattern>4</pattern><pattern_description>[Anything]</pattern_description></string_26list><int_2aint><value>2</value><default_value>2</default_value><documentation/>
<pattern>5</pattern><pattern_description>[Integer range -2147483648...2147483647 (inclusive)]</pattern_description></int_2aint><double_2bdouble><value>7.1415926</value><default_value>6.1415926</default_value><documentation>docs 3</documentation><pattern>6</pattern><pattern_description>[Double -1.79769e+308...1.79769e+308 (inclusive)]</pattern_description></double_2bdouble></Testing_25testing></ParameterHandler>
```

Reply:
Yes, exactly: the original text that had the spaces has been mangled.

Reply:
This is the original text which is read in. It is mangled because the xml writer outputs mangled text, because xml doesn't support multi word keys, and the xml ready seems to be written to only support xml which is writen out by the parameter handler.

Reply:
So what I am trying to say is that the xml read function is only tested with already mangled keys. (Because xml itself doesn't support unmangled keys)

Reply:
Yes, that's correct -- the XML reader is written with XML files in mind that were written by the XML writer, and not by hand.

You can change that for the JSON though I would suggest that you keep the JSON reader and writer in sync so that you can read files back in that the JSON writer had previously output.

Reply:
ah, I see what you mean. My preferred solution would be to always mangle and demangle like we do with the normal parameter output, but for now I will get it out and we can discuss this in a possible follow up patch.